src/checkpoint.hh

```diff
@@ -0,0 +1,1246 @@
+        c++;
+        const std::string &table_name = i->second->table_decl->name();
+        if (c == 1) {
+          stream << indent() << "std::vector<bool> *tabulated[] = {"
+                 << table_name << ".get_tabulated()," << endl;
+        } else {
+          stream << indent() << "                                  ";
+          if (c < n_tables) {
+            stream << table_name << ".get_tabulated()," << endl;
+          } else {
+            stream << table_name << ".get_tabulated()};" << endl << endl;
+          }
+        }
+      }
+      c = 0;
+      for (auto i = tables.begin(); i != tables.end(); ++i) {
+        c++;
+        const std::string &table_name = i->second->table_decl->name();
+        if (c == 1) {
+          stream << indent() << "size_t *tabulated_counts[] = {"
+                 << table_name << ".get_tabulated_count()," << endl;
+        } else {
+          stream << indent() << "                                  ";
+          if (c < n_tables) {
+            stream << table_name << ".get_tabulated_count()," << endl;
+          } else {
+            stream << table_name << ".get_tabulated_count()};" << endl << endl;
+          }
+        }
+      }
+      stream << indent() << "// create link map" << endl;
+      stream << indent() << "for (int i = 0; i < n_tables; i++) {" << endl;
+      inc_indent();
+      stream << indent() << "auto &curr_table = *(tables[i]);" << endl;
+      stream << indent() << "max_table_size = "
+             << "max(max_table_size, curr_table.size());" << endl;
+      stream << indent() << "for (size_t j = 0; "
+             << "j < curr_table.size(); j++) {" << endl;
+      inc_indent();
+      stream << indent() << "auto &l = curr_table[j].ref();" << endl;
+      stream << indent() << "for (size_t k = 0; k < l.size(); k++) {" << endl;
+      inc_indent();
+      for (size_t i = 0; i < string_type_accessors.size(); i++) {
+        stream << indent() << "if (l[k].S" << i+1 << ") {" << endl;
+        inc_indent();
+        stream << indent() << "// add address of block of current string to map"
+               << endl;
+        stream << indent() << "assert(l[k].S" << i+1
+               << " && l[k].B" << i+1 << ");" << endl;
+        stream << indent() << "String::Block *b = l[k].B" << i+1 << ";" << endl;
+        stream << indent() << "link_map[l[k].S" << i+1 << "] = b;"
+               << endl << endl;
+        stream << indent() << "unsigned char c = 0;" << endl;
+        stream << indent() << "// store table/vector idx of every linked Block"
+               << endl;
+        stream << indent() << "// so invalid links can be removed properly"
+               << endl;
+        stream << indent() << "while (c < b->pos) {" << endl;
+        inc_indent();
+        stream << indent() << "switch (b->array[c]) {" << endl;
+        inc_indent();
+        stream << indent() << "case b->REP :" << endl;
+        inc_indent();
+        stream << indent() << "c += 6;" << endl;
```

Comment:
where do these numbers like `3` or `6` stem from?

Reply:
The `+6` stems from how many characters are inserted into the block when the `REP` enum is parsed, which is 6. If you look at the `append` method in `string.hh` (line 214) you can see that 6 characters are added to the block array whenever a `REP` case is handled.
The `+3` is the equivalent for the `SEQ` case (e.g line 177 in `string.hh`). More than 3 characters can also be added to the block array in the `SEQ` case, but that is handled by the `default` switch case, which simply increments the idx by 1.

Reply:
ah, I see. Is there something like `size_of(REP)` possible, instead of hard coding these numbers?

Reply:
Hmm, I guess I could create some enum or static constant for the `+6` and `+3` cases in the `String` class, but that won't really connect these numbers to the internals of the `append` methods, so that would be kind of useless. Also I only need these values in the `restore_string_links` method, so it wouldn't make much sense to add them to the `String` class if they aren't even going to be needed in any of the class methods. As long as you're not planning to change the `String` internals anytime soon I would suggest to simply keep these numbers hard-coded.

Reply:
ok, keep as is!